### ISSUE TITLE
Universal predicates

### ISSUE THREAD

**Issue by grigandal625** [antlr/antlr4#4347]
Is it planned to implement a universal way to specify predicates that does not depend on the target language of the generated parser?

**Comment by ericvergnaud** [comment]
See #4345Envoyé de mon iPhoneLe 5 juil. 2023 à 22:17, Андрей Григорьев ***@***.***> a écrit :﻿
Is it planned to implement a universal way to specify predicates that does not depend on the target language of the generated parser?

**Comment by DavidGregory084** [comment]
@grigandal625 I've had a go at enabling something like that in #4345. There's also been a lot of discussion about this in #4067.

**Comment by kaby76** [comment]
It isn't a "universal language predicate", but https://github.com/antlr/antlr4/pull/4345 goes a way in fixing the problem of forking a grammar for each target. You still have to write target-specific code, which can be pretty challenging. I am still trying the PR out on the [python3 grammar](https://github.com/antlr/grammars-v4/pull/3540), but [it looks pretty good](https://github.com/kaby76/stg-python3) so far with Cpp, CSharp, Java, Go (with Go runtime fix) working.

**Comment by ericvergnaud** [comment]
Yes it’s promising. We’ll see if it makes it to merge.Envoyé de mon iPhoneLe 8 juil. 2023 à 15:25, Ken Domino ***@***.***> a écrit :﻿
It isn't a "universal language predicate", but #4345 goes a way in fixing the problem of forking a grammar for each target. You still have to write target-specific code, which can be pretty challenging. I am still trying the PR out on the python3 grammar, but it looks pretty good so far with Cpp, CSharp, Java, Go (with Go runtime fix) working.
